Implement CS.R.44

Javier-varez opened the pull request.

Javier-varez committed:
Implement CS.R.44

finomen commented on clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp:
```diff
@@ -53,6 +57,173 @@ void TypeConversionsCheck::check(const MatchFinder::MatchResult &Result) {
+      return true;
+    }
+
+    // If they are the same sign, they need to have at least the same number of
```
Promotion has more strict definition

Javier-varez replied:
You are right, it is more strict, but these conversions are safe, so I see no reason not to allow them.

Javier-varez replied:
Renamed to `isSafeIntegralCast` to make it clear it is not only promotion

finomen commented on clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp:
```diff
@@ -53,6 +57,173 @@ void TypeConversionsCheck::check(const MatchFinder::MatchResult &Result) {
+  return false;
+}
+
+bool TypeConversionsCheck::isValidFloatingPointPromotion(
```
There is no such thing as floating point promotion

Javier-varez replied:
I know, but I think this should make it easier to deal with these warnings, since the conversions this checks allows are safe. If you prefer, I can remove it, at the cost of always making the user call clipToFit even for larger types.

Javier-varez replied:
Removed

aburgm commented on clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp:
```diff
@@ -53,6 +57,140 @@ void TypeConversionsCheck::check(const MatchFinder::MatchResult &Result) {
+    }
+  }
+
+  if (isSafeIntegralCast(context, sourceType, destType)) {
```
The rule in the CS says "All type conversions MUST be explicit.". So in principle this also applies to safe integral casts. But maybe we can relax the rule to say "all type conversions MUST be explicit or not lead to data loss" (instead of and)...

Javier-varez replied:
Yes, I discussed this with Nikolai yesterday, I think it  would make more sense to make our lives a bit easier. This change also allows implicit casts that would be extremely noisy like:
- Adding qualifiers to a pointed type.
- Array to pointer decay. (which I realized is currently broken).
- Function to function pointer decay (which clang seems to do always before calling a function).
- explicit casts where source and destination are the same (which can happen due to integral promotion and then a static_cast).

aburgm replied:
Sounds good to me.

Javier-varez replied:
Also added here:
 - Dependent types, which should be checked in instantiation.
 - Removing qualifiers by value implicitly (which is always safe)

aburgm commented on clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp:
```diff
@@ -53,6 +57,140 @@ void TypeConversionsCheck::check(const MatchFinder::MatchResult &Result) {
+  }
+
+  if (sourceType->isIntegerType() && destType->isRealFloatingType()) {
+    // Allowed integral to floating point type
```
although in principle there could be data loss here

Javier-varez replied:
That is a fair  point indeed, but we should be able to make this conversion explicitly. Or would you prefer using a explicit function for these conversions (similar to `clipToFit<T>`)? I also understand that these methods would internally just use static cast and have a exception for this rule.

aburgm replied:
static_cast is fine, but it means the "no data loss" check is not going to be fully automated and still needs to be checked for during review. Right?

Javier-varez replied:
I would think so, yes. Or we can remove this and force the user to explain themselves by adding an explicit exception?

aburgm replied:
Either works for me

Javier-varez replied:
Went for the second option, so removed the exception from integral to floating point, users will need `// NOLINT` for that conversion or clipToFit or similar methods.

Javier-varez committed:
Exceptions to type-conversions

-Pointer types to void*
-nullptr to pointer types
-builtin functions to function pointer

Javier-varez commented:
@finomen  Could you have one more look at this? I think it should be fine now.

Added 3 more exceptions:
- Builtin fn type to pointer type for pointer decay.
- nullptr to pointer type
- pointer  type to void *.
